asdf/commands/edit.py

```diff
@@ -0,0 +1,589 @@
+    fd.write(bindex_hdr)
+    for idx in index:
+        ostr = f"- {idx}\n"
+        fd.write(ostr.encode("utf-8"))
+    end = b"..."
+    fd.write(end)
+    return
+
+
+def get_next_block_header(fd):
+    """
+    From a file, gets the next block header.
+
+    Parameters
+    ----------
+    fd - The ASDF file to get the next block.
+
+    Return
+    ------
+    If a block is found, return the bytes of the block header.
+    Otherwise return None.
+    """
+    #     Block header structure:
+    # 4 bytes of magic number
+    # 2 bytes of header length, after the length field (min 48)
+    # 4 bytes flag
+    # 4 bytes compression
+    # 8 bytes allocated size
+    # 8 bytes used (on disk) size
+    # 8 bytes data size
+    # 16 bytes checksum
+    blk_header = fd.read(6)
+    if len(blk_header) != 6:
+        return None
+    if not blk_header.startswith(constants.BLOCK_MAGIC):
+        return None
+    hsz = struct.unpack(">H", blk_header[4:6])[0]
+    header = fd.read(hsz)
+    return blk_header + header
+
+
+def rewrite_asdf_file(edited_text, orig_text, oname, fname):
+    """
+    Rewrite an ASDF file for too large edited YAML.  The edited YAML, a pad,
+    the blocks will be rewritten.  A block index will also be rewritten.  If a
+    block index existed in the old file, it will have to be recomputed to
+    because of the larger YAML size and pad, which changes the location of
+    the binary blocks.
+
+    Parameters
+    ----------
+    edited_text : the new YAML text to write out.
+    orig_text : the original YAML text to overwrite.
+    oname : the ASDF file to overwrite.
+    fname : the edit YAML to write to new file.
+    """
+
+    tmp_oname = oname + ".tmp"  # Save as a temp file, in case anything goes wrong.
+    pad_size = 10 * 1000
+    padded_text = add_pad_to_new_text(edited_text, pad_size)
+
+    ifd = open(oname, "r+b")  # Open old ASDF to get binary blocks
+    ifd.seek(len(orig_text))
+
+    ofd = open(tmp_oname, "w+b")  # Open temp file to write
+    ofd.write(padded_text)  # Write edited YAML
+
+    current_location = len(padded_text)
+    block_index = []
+    alloc_loc = 14  # 4 bytes of block ID, 2 blocks of size, 8 blocks into header
+    block_chunk = 2048
+    while True:
+        next_block = get_next_block_header(ifd)
+        if next_block is None:
+            break
+
+        # Get block size on disk
+        alloc = struct.unpack(">Q", next_block[alloc_loc : alloc_loc + 8])[0]
+
+        # Save block location for block index
+        block_index.append(current_location)
+        current_location = current_location + len(next_block) + alloc
+
+        # Copy block
+        ofd.write(next_block)
+        while alloc >= block_chunk:
+            chunk = ifd.read(block_chunk)
+            ofd.write(chunk)
+            alloc -= block_chunk
+        if alloc > 0:
+            chunk = ifd.read(alloc)
+            ofd.write(chunk)
+    ifd.close()
+
+    write_block_index(ofd, block_index)
```

Comment:
I mentioned this in my previous review, but this is not correct for files that contain a streaming block.  Per the standard the block index should not be written when a file utilizes the streaming block.

Reply:
I have added the case for a streamed binary block and included a test for it.